index.bs

```diff
@@ -47,66 +51,88 @@ dictionary CSSStyleSheetInit {
-		3. If the {{CSSStyleSheetInit/alternate}} attribute of <var>options</var> is true,
+		4. If the {{CSSStyleSheetInit/alternate}} attribute of <var>options</var> is true,
 			set <var>sheet’s</var> alternate flag.
-		4. If the {{CSSStyleSheetInit/disabled}} attribute of <var>options</var> is true,
+		5. If the {{CSSStyleSheetInit/disabled}} attribute of <var>options</var> is true,
 			set <var>sheet’s</var> disabled flag.
-		5. <a spec=css-syntax-3>Parse a stylesheet</a> from {{text}}.
+		6. Return <var>sheet</var>.
+	</dd>
+	<dt><dfn  for=CSSStyleSheet lt="constructed flag">constructed flag</dfn></dt>
+	<dd>
+		Specified when created. Either set or unset. Unset by default.
+		Signifies whether this stylesheet is made via constructor or not, so must be set only for stylesheets that are constructed using the {{CSSStyleSheet()}} or {{CSSStyleSheet(options)}} function.
+	</dd>
+	<dt><dfn  for=CSSStyleSheet lt="disallow modification flag">disallow modification flag</dfn></dt>
+	<dd>
+		Either set or unset. Unset by default. If set, modification to the stylesheet's rules are not allowed.
+	</dd>
+	<dt><dfn  for=CSSStyleSheet lt="constructor document">constructor document</dfn></dt>
+	<dd>
+		Specified when created. The {{Document}} where the CSSStyleSheet is originally constructed on. Null by default.
+	</dd>
+	<dt><dfn  for=CSSStyleSheet lt="set of adopter documents">set of adopter documents</dfn></dt>
+	<dd>
+		A set of {{Document}}s on which the stylesheet is adopted (e.g. there is at least one {{DocumentOrShadowRoot}} within that {{Document}}'s tree on which the sheet is part of the {{DocumentOrShadowRoot}}'s {{DocumentOrShadowRoot/adoptedStyleSheets}}).
+	</dd>
+</dl>
+
+Modifying Constructed Stylesheets {#modifying-constructed-stylesheets}
+=============================
+
+After construction, constructed stylesheets can be modified using rule modification methods like <a spec="cssom-1" for="CSSStyleSheet">insertRule(rule)</a> or <a spec="cssom-1" for="CSSStyleSheet">deleteRule(rule)</a>, or {{replace(text)}} and {{replaceSync(text)}} if the sheet's [=disallow modification flag=] is not set. If those methods are called when the sheet's [=disallow modification flag=] is set, or <a spec="cssom-1" for="CSSStyleSheet">insertRule(rule)</a> is used to add an <a spec=css-cascade-4>@import</a> rule, a "{{NotAllowedError}}" {{DOMException}} must be thrown. 
+
+<dl>
+    <dt><dfn method for=CSSStyleSheet lt="replace(text)">replace(text)</dfn></dt>
+    <dd>
+		1. Let <var>sheet</var> be the stylesheet on which this function is called on.
+		2. If <var>sheet’s</var> [=constructed flag=] is not set, or <var>sheet's</var> [=disallow modification flag=] is set, throw a "{{NotAllowedError}}" {{DOMException}}.
+		3. Set <var>sheet’s</var> CSS rules to an empty list, and set <var>sheet's</var> [=disallow modification flag=].
+		4. <a spec=css-syntax-3>Parse a stylesheet</a> from {{text}}.
```

Comment:
It looks like this doesn't return a list of rules; instead it returns a stylesheet of some sort. (But probably not the same as a CSSOM stylesheet?) This will probably need to be refactored eventually, but that's more of a @tabatkins issue. It doesn't need to block this, but we may want to file a tracking issue for this.

Reply:
Looks like there's https://drafts.csswg.org/css-syntax-3/#parse-list-of-rules, used that now.

Reply:
That @charset handling is just so that other rules that check for "is this the first rule in the stylesheet?" don't fail because of a charset indicator.

That said, you bring up a good point - I should make the parsing algorithms *optionally* just allow themselves to be parsed in a forced utf-8, and then we should force that here. No reason to support other charsets; that's a legacy thing.

The *other* difference is whether `<!--` and `-->` are ignored during parsing or not. This is another *very* legacy thing, to support commenting out the contents of `style` blocks to hide them from browsers that didn't understand CSS and thus treated `style` as an unknown element.  This is also irrelevant for modern times, and so it's fine to not support it.

So all in all, I'm fine with switching to this algorithm.

Reply:
> and then we should force that here

I'm not following.   The input here is a DOMString, not bytes.  What does it mean to parse that "as utf-8"?  Are we expecting people to pass in strings or bytes here?  Is there value in having this function take USVString, or do we not have a CSS parsing  algorithm that takes a Unicode string as input?   If we're supposed to be passing bytes, we should make that explicit by making this a ByteString, but I expect that's not what people actually want here.

Reply:
Ah, so, the input to the tokenizer is a stream of unicode codepoints (per <https://drafts.csswg.org/css-syntax/#tokenization>).  The spec also specifies (in [3.2](https://drafts.csswg.org/css-syntax/#input-byte-stream)) that if you're actually starting from a stream of bytes, there are some rules for how to transform it into a stream of codepoints.  (This series of transformations is more explicitly laid out in one spot in [5.3](https://drafts.csswg.org/css-syntax/#parser-entry-points)

That all said, DOMString doesn't actually contain codepoints. :/

Reply:
Does the tokenize take a stream of codepoints or scalar values?

Treating DOMString as a stream of codepoints is doable.   But if you want scalar values, and I suspect you do, USVString is probably a better fit.

Reply:
Just codepoints, tho I suspect I did actually mean scalar values. (Back when I wrote this I don't think we were making the distinction clearly in DOM.) The fact that you can't use an escape to produce a surrogate suggests that I'm assuming no lone surrogates would show up in the input stream either.

Reply:
I don't really understand the conversation, but I take that it's OK to use the "parse list of rules" because the charset and comments thing are legacy and OK to not be supported?

Reply:
Yes. There might be a little detail I need to fix up in Syntax to handle (filed as <https://github.com/w3c/csswg-drafts/issues/3307>), but that either won't need you to do anything, or will just require a small addition that I'll ping you about.

Reply:
Cool! Do you think the rest of this PR is ok to merge?